### ISSUE TITLE
execution_requirements keys are filtered like tags

### ISSUE THREAD

**Issue by dws** [bazelbuild/bazel#24616]
### Description of the bug:

The implementation of `--incompatible_allow_tags_propagation` deliberately only propagates certain tags to the `ExecutionInfo` of an action.  The filtering on tags appears to be accomplished via `legalExecInfoKeys()` in `src/main/java/com/google/devtools/build/lib/packages/TargetUtils.java`.

The surprise is that the keys of the `execution_requirements` dict parameter to `ctx.actions.run()` and `cts.actions.run_shell()` are also filtered in this way.

By contrast, the keys of the implicit rule attribute `exec_properties` are not filtered.

I can understand wanting to limit what `tags` propagate into `ExecutionInfo`.  But since `execution_requirements` is explicitly there only for augmenting `ExecutionInfo` it is most surprising and counterintuitive that its keys would be filtered, especially because the keys of `exec_properties` are not filtered.

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

With the following `BUILD.bazel`:
```
load(":repro.bzl", "repro_run", "repro_run_shell")

# The tags are filtered by legalExecInfoKeys(), so only
# "supports-repro-tag" makes it into ExecutionInfo, as intended.

REPRO_TAGS = [
    "RENEGADE-repro-tag",       # does not satisfy legalExecInfoKeys()
    "supports-repro-tag",       # satisfies legalExecInfoKeys()
]

# Both entries in this exec_properties dict make it into ExecutionInfo,
# as expected.

REPRO_EXEC_PROPERTIES = {
    "RENEGADE-repro-exec-property": "", # does not satisfy legalExecInfoKeys()
    "supports-repro-exec-property": "", # satisfies legalExecInfoKeys()
}

# The surprise is that the keys of this execution_requirements
# dict get filtered by legalExecInfoKeys(), so only
# "supports-repro-execution-requirement" makes it into ExecutionInfo.

REPRO_EXECUTION_REQUIREMENTS = {
    "RENEGADE-repro-execution-requirement": "", # does not satisfy legalExecInfoKeys()
    "supports-repro-execution-requirement": "", # satisfies legalExecInfoKeys()
}

repro_run(
    name = "repro-run",
    out = "repro-run.out",
    tags = REPRO_TAGS,
    exec_properties = REPRO_EXEC_PROPERTIES,
    execution_requirements = REPRO_EXECUTION_REQUIREMENTS,
)

repro_run_shell(
    name = "repro-run-shell",
    out = "repro-run-shell.out",
    tags = REPRO_TAGS,
    exec_properties = REPRO_EXEC_PROPERTIES,
    execution_requirements = REPRO_EXECUTION_REQUIREMENTS,
)
```
and the following `repro.bzl`:
```
def _repro_run_impl(ctx):

    output = ctx.outputs.out

    ctx.actions.run(
        outputs = [output],
        executable = "/usr/bin/touch",
        arguments = [output.path],
        execution_requirements = ctx.attr.execution_requirements,
    )

    return DefaultInfo(files = depset([output]))

repro_run = rule(
    implementation = _repro_run_impl,
    attrs = {
        "out": attr.output(),
        "execution_requirements": attr.string_dict(),
    },
)

def _repro_run_shell_impl(ctx):

    output = ctx.outputs.out

    ctx.actions.run_shell(
        outputs = [output],
        command = "/usr/bin/touch " + output.path,
        execution_requirements = ctx.attr.execution_requirements,
    )

    return DefaultInfo(files = depset([output]))


repro_run_shell = rule(
    implementation = _repro_run_shell_impl,
    attrs = {
        "out": attr.output(),
        "execution_requirements": attr.string_dict(),
    },
)
```
and an empty `MODULE.bazel` file, I see the following:
```
$ bazel --quiet aquery //:all | grep ExecutionInfo
  ExecutionInfo: {RENEGADE-repro-exec-property: '', supports-repro-exec-property: '', supports-repro-execution-requirement: '', supports-repro-tag: ''}
  ExecutionInfo: {RENEGADE-repro-exec-property: '', supports-repro-exec-property: '', supports-repro-execution-requirement: '', supports-repro-tag: ''}
```
but I would expect to see the following:
```
$ bazel --quiet aquery //:all | grep ExecutionInfo
    ExecutionInfo: {RENEGADE-repro-exec-property: '', RENEGADE-repro-execution-requirement: '', supports-repro-exec-property: '', supports-repro-execution-requirement: '', supports-repro-tag: ''}
  ExecutionInfo: {RENEGADE-repro-exec-property: '', RENEGADE-repro-execution-requirement: '', supports-repro-exec-property: '', supports-repro-execution-requirement: '', supports-repro-tag: ''}
```
In other words, I would expect to see `legalExecInfoKeys()` apply only to
`tags`, and not to the keys of the `execution_requirements` parameter to
`ctx.actions.run()` or `ctx.actions.run_shell()`.

### Which operating system are you running Bazel on?

Ubuntu 20.04

### What is the output of `bazel info release`?

release 8.0.0

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.

This also occurs in Bazel 7.2.1.  I did not try to see just how far back this behavior goes.

### Have you found anything relevant by searching the web?

No

**Comment by fmeum** [comment]
Does this also happen with the propagation flag unflipped?

**Comment by dws** [comment]
> Does this also happen with the propagation flag unflipped?

Yes.

With `--noincompatible_allow_tags_propagation` the `tags` no longer get added to the `ExecutionInfo`, but the `execution_requirements` dict keys are still filtered.
```
$ ../bazel-8.0.0-linux-x86_64 --quiet aquery --noincompatible_allow_tags_propagation //:all | grep ExecutionInfo
  ExecutionInfo: {RENEGADE-repro-exec-property: '', supports-repro-exec-property: '', supports-repro-execution-requirement: ''}
  ExecutionInfo: {RENEGADE-repro-exec-property: '', supports-repro-exec-property: '', supports-repro-execution-requirement: ''}
```

**Comment by fmeum** [comment]
Looks like this was an explicit choice: https://github.com/bazelbuild/bazel/commit/5b7909853029b0ffaeea7336b0abdf1557d71802

The test case shows how certain execution infos should reasonably be blocked (e.g. `timeout`). Being able to set it via `exec_properties` also sounds like something that should be prohibited in this case, I don't know why that's not done.

@kotlaja Can you perhaps shed some light on this?

@dws What's the actual use case in which you would want to set an execution info key that is currently being filtered out?

**Comment by dws** [comment]
> Looks like this was an explicit choice: [5b79098](https://github.com/bazelbuild/bazel/commit/5b7909853029b0ffaeea7336b0abdf1557d71802)
> 
> The test case shows how certain execution infos should reasonably be blocked (e.g. `timeout`). Being able to set it via `exec_properties` also sounds like something that should be prohibited in this case, I don't know why that's not done.
> 
> @kotlaja Can you perhaps shed some light on this?
> 
> @dws What's the actual use case in which you would want to set an execution info key that is currently being filtered out?

This behavior came to light due to a confluence of two circumstances:
1) We had a custom rule with no macro wrapper that was naively expecting to be able to pass a dict suitable for `exec_properties` to `execution_requirements` on the action it was generating.
2) We need for `ExecutionInfo` on this rule to contain the key `gpu` since our buildfarm setup uses that to direct the action to a particular pool of runners.

But, if we put `gpu` in `execution_requrements`, it gets filtered out silently and never makes it to the remote execution system.

I could understand the filtering if Bazel itself was the only thing interpreting the `ExecutionInfo`, but that is simply not the case.
